lib/grandpa/grandpa.go

```diff
@@ -1281,3 +1167,209 @@ func (s *Service) lenVotes(stage Subround) int {
+	logger.Debugf("received vote message, (peer: %s, msg: %+v)", from, vote)
+	s.sendTelemetryVoteMessage(vote)
+
+	v, err := s.validateVoteMessage(from, vote)
+	if err != nil {
+		// TODO: log in the outer most layer and maybe split the erros from the return error string
+		logger.Debugf("failed to validate vote message {%v}: %s", vote, err)
+		return
+	}
+
+	logger.Debugf(
+		"validated vote message %v from %s, round %d, subround %d, "+
+			"prevote count %d, precommit count %d, votes needed %d",
+		v, vote.Message.AuthorityID, vote.Round, vote.Message.Stage,
+		s.lenVotes(prevote), s.lenVotes(precommit), s.state.threshold()+1)
+}
+
+func (s *Service) handleCommitMessage(commitMessage *CommitMessage) error {
+	logger.Debugf("received commit message, msg: %+v", commitMessage)
+
+	err := verifyBlockHashAgainstBlockNumber(s.blockState,
+		commitMessage.Vote.Hash, uint(commitMessage.Vote.Number))
+	if err != nil {
+		if errors.Is(err, chaindb.ErrKeyNotFound) {
+			s.tracker.addCommit(commitMessage)
+			logger.Infof("we might not have synced to the given block %s yet: %s",
+				commitMessage.Vote.Hash, err)
+			return nil
+		}
+
+		return fmt.Errorf("verifying block hash against block number: %w", err)
+	}
+
+	containsPrecommitsSignedBy := make([]string, len(commitMessage.AuthData))
+	for i, authData := range commitMessage.AuthData {
+		containsPrecommitsSignedBy[i] = authData.AuthorityID.String()
+	}
+
+	s.telemetry.SendMessage(
+		telemetry.NewAfgReceivedCommit(
+			commitMessage.Vote.Hash,
+			fmt.Sprint(commitMessage.Vote.Number),
+			containsPrecommitsSignedBy,
+		),
+	)
+
+	has, err := s.blockState.HasFinalisedBlock(commitMessage.Round, s.state.setID)
+	if err != nil {
+		return fmt.Errorf("checking for a finalized block in the block state: %w", err)
+	}
+
+	if has {
+		return nil
+	}
+
+	// check justification here
+	err = verifyCommitMessageJustification(commitMessage, s.state.setID,
+		s.state.threshold(), s.authorities(), s.blockState)
+	if err != nil {
+		if errors.Is(err, blocktree.ErrStartNodeNotFound) {
+			// we haven't synced the committed block yet, add this to the tracker for later processing
+			s.tracker.addCommit(commitMessage)
+		}
+		return fmt.Errorf("verifying commit message justification: %w", err)
+	}
+
+	err = s.blockState.SetFinalisedHash(commitMessage.Vote.Hash, commitMessage.Round, s.state.setID)
+	if err != nil {
+		return fmt.Errorf("setting finalised hash: %w", err)
+	}
+
+	preCommitSigned, err := compactToJustification(commitMessage.Precommits, commitMessage.AuthData)
+	if err != nil {
+		return fmt.Errorf("compacting justification: %w", err)
+	}
+
+	err = s.grandpaState.SetPrecommits(commitMessage.Round, commitMessage.SetID, preCommitSigned)
+	if err != nil {
+		return fmt.Errorf("setting precommits: %w", err)
+	}
+
+	// TODO: re-add catch-up logic (#1531)
+	return nil
+}
+
+func verifyCommitMessageJustification(commitMessage *CommitMessage, setID uint64, threshold uint64,
```

Comment:
nit pass commitMessage as `CommitMessage` without a pointer, I don't think there is a need :thinking:

Reply:
It will require more deltas that are not in the scope of this PR (this PR is quite large), we could open a new issue if you think we should really address it otherwise I think there is no problem to keep it as it is